modules/javafx.graphics/src/jslc/java/com/sun/scenario/effect/compiler/JSLC.java

```diff
@@ -199,7 +208,8 @@ private static ParserInfo compile(JSLCInfo jslcinfo,
             File outFile = jslcinfo.getOutputFile(OUT_D3D);
             if (jslcinfo.force || outOfDate(outFile, sourceTime)) {
                 if (pinfo == null) pinfo = getParserInfo(stream);
-                HLSLBackend hlslBackend = new HLSLBackend(pinfo.parser, pinfo.program);
+                HLSLBackend hlslBackend = new HLSLBackend(pinfo.parser, pinfo.visitor);
+                hlslBackend.scan(pinfo.program);
```

Comment:
I'm trying to remember why I did that. I think it makes sense to change them to all be how I did GLSL and HLSL because it removes "side effects" and seems cleaner. I don't think I did them differently for any good reason. Do you have a preference as to changing them all to be like HLSL/GLSL or reverting how I changed HLSL/GLSL backends?

Reply:
Looking at it again, I think that because the ProgramUnit parameter is never actually used inside the class it makes sense to take it out of the constructor and only pass it along to `scan`.

Reply:
Looking at it again I see that for HLSLBackend ans ES2Backend that they, unlike the other backends, do not contain `private final String body` members. Therefore for HLSLBackend and ES2Backend the `ProgramUnit` parameter was only used to pass to `scanner.scan(program);`. In the other backends `body` is assigned thusly `this.body = scanner.getResult();` and then `getGenCode` is called for `MEBackend`, `SSEBackend`, and `JSWBackend` which returns a `String` which is then written to a file. For `HLSLBackend` and `ES2Backend`, because they extend `SLBackend`, they work in a different way. Namely they have a `private StringBuilder sb = new StringBuilder();` member which is set to the result of the generated shader (instead of `getGenCode` returning the shader as a `String`). I hope that makes sense, it's a bit confusing actually. The tldr is that I was able to easily cleanup HLSL and ES2 backends but the others, because they are structured differently, would have required much more restructuring for little benefit (IMHO).

Reply:
OK, thanks for the explanation. I think it is fine, then.